cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -87,6 +89,195 @@ Status GetListElementIndex(const ExecValue& value, T* out) {
   return Status::OK();
 }
 
+template <typename Type, typename IndexType>
+struct ListSlice {
+  using offset_type = typename Type::offset_type;
+
+  static Status Exec(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+    const auto opts = OptionsWrapper<ListSliceOptions>::Get(ctx);
+
+    // Invariants
+    if (!opts.stop.has_value()) {
+      // TODO: Support slicing to arbitrary end
+      // For variable size list, this would be the largest difference in offsets
+      // For fixed size list, this would be the fixed size.
+      return Status::NotImplemented(
+          "Slicing to end not yet implemented, please set `stop` parameter.");
+    }
+    if (opts.start < 0 || opts.start >= opts.stop.value()) {
+      // TODO: support start == stop which should give empty lists
+      return Status::Invalid("`start`(", opts.start,
+                             ") should be greater than 0 and smaller than `stop`(",
+                             opts.stop, ")");
+    }
+    if (opts.step != 1) {
+      // TODO: support step in slicing
+      return Status::NotImplemented(
+          "Setting `step` to anything other than 1 is not supported; got step=",
+          opts.step);
+    }
+
+    const ArraySpan& list_ = batch[0].array;
+    const Type* list_type = checked_cast<const Type*>(list_.type);
+    const auto value_type = list_type->value_type();
+
+    std::unique_ptr<ArrayBuilder> builder;
+
+    // construct array values
+    if (opts.return_fixed_size_list) {
+      RETURN_NOT_OK(MakeBuilder(
+          ctx->memory_pool(),
+          fixed_size_list(value_type,
+                          static_cast<int32_t>(opts.stop.value() - opts.start)),
+          &builder));
+      RETURN_NOT_OK(BuildArray<FixedSizeListBuilder>(batch, opts, *builder));
+    } else {
+      if constexpr (std::is_same_v<Type, LargeListType>) {
+        RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), large_list(value_type), &builder));
+        RETURN_NOT_OK(BuildArray<LargeListBuilder>(batch, opts, *builder));
+      } else {
+        RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), list(value_type), &builder));
```

Comment:
Instead of creating a (large_)list type, can we use the original one (`list_type` defined above)? That would then preserve the field names?

Reply:
As mentioned in https://github.com/apache/arrow/pull/14395/commits/8a3bc6d2e7a3f051bad2b26536711d8a89b10e25, it seems, naively, the field names are preserved?

Reply:
See my comment on the test you added, it's the name of the list child field itself we need to test (and I think that will still require code changes to get that working)